src/crimson/osd/main_config_bootstrap_helpers.cc

```diff
@@ -143,43 +161,79 @@ _get_early_config(int argc, const char *argv[])
 	  std::begin(early_args),
 	  std::end(early_args));
 
-	if (auto found = std::find_if(
-	      std::begin(early_args),
-	      std::end(early_args),
-	      [](auto* arg) { return "--cpuset"sv == arg; });
-	    found == std::end(early_args)) {
-	  auto cpu_cores = crimson::common::get_conf<std::string>("crimson_seastar_cpu_cores");
-	  if (!cpu_cores.empty()) {
-	    // Set --cpuset based on crimson_seastar_cpu_cores config option
-	    // --smp default is one per CPU
-	    ret.early_args.emplace_back("--cpuset");
-	    ret.early_args.emplace_back(cpu_cores);
-	    ret.early_args.emplace_back("--thread-affinity");
-	    ret.early_args.emplace_back("1");
-	    logger().info("get_early_config: set --thread-affinity 1 --cpuset {}",
-	                  cpu_cores);
-	  } else {
-	    auto reactor_num = crimson::common::get_conf<uint64_t>("crimson_seastar_num_threads");
-	    if (!reactor_num) {
-	      logger().error("get_early_config: crimson_seastar_cpu_cores"
+        auto found = std::find_if(
+              std::begin(early_args),
+              std::end(early_args),
+              [](auto* arg) { return "--cpuset"sv == arg; });
+
+        bool cpuset_set = false;
+        bool smp_set = false;
+        std::optional<OptionValue> tmp_value;
+        for (const auto& option : seastar_options) {
+           std::optional<OptionValue> option_value;
+           if ((found == std::end(early_args)) && option.option_name == "--cpuset") {
+             tmp_value = crimson::common::get_conf<std::string>("crimson_seastar_cpu_cores");
+             if (std::holds_alternative<std::string>(tmp_value.value())) {
+               const std::string& str_value = std::get<std::string>(tmp_value.value());
+               if (str_value != "0-0") {
+                 cpuset_set = true;
+                 option_value = tmp_value;
+               }
+             }
+           }
+           if ((found == std::end(early_args)) && option.option_name == "--smp" && !cpuset_set) {
```

Comment:
Perhaps this is a good opportunity to re-structure the logic here. It makes it hard to follow the more nuances and more options are supported. Suggesting the following structure, any other ideas are welcomed:

* Verify that no direct Seastar options were passed (without the conf) into early_args as this is prohibited.
* Move the CPU strategy selected (`crimson_seastar_cpu_cores` or `crimson_seastar_num_threads`) out of the common option flow.
* Introduce a generalized "trivial" option emplacement logic:
  * Iterate through `seastar_options` and emplace each option accordingly
  * Additionally, print each option that is being set, same as the log lines we already had.

Reply:
done